packages/core/src/shader/UniformGroup.ts

```diff
@@ -1,36 +1,73 @@
+import { BUFFER_TYPE } from '@pixi/constants/src';
 import type { Dict } from '@pixi/utils';
+import { Buffer } from '../geometry/Buffer';
 import type { UniformsSyncCallback } from './utils';
 
 let UID = 0;
 
 /**
  * Uniform group holds uniform map and some ID's for work
  *
+ * The uniform group has two modes:
+ *
+ * 1: normal mode
+ * Normal mode will upload the uniforms with individual function calls as required
+ *
+ * 2: uniform buffer mode
+ * This mode will treat the uniforms as a uniform buffer. You can pass in either a buffer that you manually handle, or
+ * or a generic object that pixi will automatically map to a buffer for you.
+ * For maximum benefits, its best to make Ubo UniformGroups static, and only update them each frame.
+ *
+ * rules of UBO's:
+ * - ubo's only work with WebGL2, so make sure you have a fallback!
+ * - only floats are supported (including vec[2,3,4], mat[2,3,4])
+ * - samplers cannot be used i ubo's (a GPU limitation)
+ * - you must ensure that the object you pass in exactly matches in the shader ubo structure.
+ * Otherwise, weirdness will ensue!
+ * - the name of the ubo object added to the group must match exactly the name of the ubo in the shader.
+ *
+ * ```
+ * // ubo in shader:
+ * uniform myCoolData { // declaring a ubo..
+ * mat4 uCoolMatrix;
+ * float uFloatyMcFloatFace
+ *
+ *
+ * // a new uniform buffer object..
+ * const myCoolData = new UniformBufferGroup({
+ *   uCoolMatrix: new Matrix(),
+ *   uFloatyMcFloatFace: 23,
+ * }}
+ *
+ * // build a shader...
+ * const shader = Shader.from(srcVert, srcFrag, {
+ *   myCoolData // name matches the ubo name in the shader. will be processed accordingly.
+ * })
+ *
+ *  ```
+ *
  * @class
  * @memberof PIXI
  */
-export class UniformGroup
+export class UniformGroup<LAYOUT = Dict<any>>
 {
-    public readonly uniforms: Dict<any>;
+    public readonly uniforms: LAYOUT;
     public readonly group: boolean;
     public id: number;
     syncUniforms: Dict<UniformsSyncCallback>;
     dirtyId: number;
     static: boolean;
+    ubo: boolean;
+    buffer?: Buffer;
+    autoManage: boolean;
 
     /**
-     * @param {object} [uniforms] - Custom uniforms to use to augment the built-in ones.
+     * @param {object | Buffer} [uniforms] - Custom uniforms to use to augment the built-in ones. Or a pixi buffer
      * @param {boolean} [_static] - Uniforms wont be changed after creation
+     * @param {boolean} [_ubo] - if true, will treat this uniform group as a uniform buffer object
      */
-    constructor(uniforms: Dict<any>, _static?: boolean)
+    constructor(uniforms: LAYOUT | Buffer, _static?: boolean, _ubo?: boolean)
```

Comment:
Why do `_static` and `_ubo` have a `_` prefix?

Reply:
possible cos they are optional? 
i'm not really sure :D

Reply:
ohh! static is a reserved key word :D